x/mongo/driver/topology/pool.go

```diff
@@ -601,3 +517,392 @@ func (p *pool) clear(serviceID *primitive.ObjectID) {
+			p.idleConns = append(p.idleConns, conn)
+		}
+
+		// If we got here, we tried to deliver an idle conn to w. No matter if tryDeliver() returned
+		// true or false, w is no longer waiting and doesn't need to be added to any wait queues, so
+		// return delivered = true.
+		return true
+	}
+
+	p.idleConnWait.cleanFront()
+	p.idleConnWait.pushBack(w)
+	return false
+}
+
+func (p *pool) queueForNewConn(w *wantConn) {
+	p.connsCond.L.Lock()
+	defer p.connsCond.L.Unlock()
+
+	p.newConnWait.cleanFront()
+	p.newConnWait.pushBack(w)
+	p.connsCond.Signal()
+}
+
+func (p *pool) totalConnectionCount() int {
+	p.connsCond.L.Lock()
+	defer p.connsCond.L.Unlock()
+
+	return len(p.conns)
+}
+
+func (p *pool) availableConnectionCount() int {
+	p.idleMu.Lock()
+	defer p.idleMu.Unlock()
+
+	return len(p.idleConns)
+}
+
+// createConnections creates connections for wantConn requests on the newConnWait queue.
+func (p *pool) createConnections(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	// condition returns true if the createConnections() loop should continue and false if it should
+	// wait. Note that the condition also listens for Context cancellation, which also causes the
+	// loop to continue, allowing for a subsequent check to return from createConnections().
+	condition := func() bool {
+		checkOutWaiting := p.newConnWait.len() > 0
+		poolHasSpace := p.maxSize == 0 || uint64(len(p.conns)) < p.maxSize
+		cancelled := ctx.Err() != nil
+		return (checkOutWaiting && poolHasSpace) || cancelled
+	}
+
+	// wait waits for there to be an available wantConn and for the pool to have space for a new
+	// connection. When the condition becomes true, it creates a new connection and returns the
+	// waiting waiting wantConn and new connection. If the Context is cancelled or there are any
+	// errrors, wait returns with "ok = false".
+	wait := func() (*wantConn, *connection, bool) {
+		p.connsCond.L.Lock()
+		defer p.connsCond.L.Unlock()
+
+		for !condition() {
+			p.connsCond.Wait()
+		}
+
+		if ctx.Err() != nil {
+			return nil, nil, false
+		}
+
+		p.newConnWait.cleanFront()
+		w := p.newConnWait.popFront()
+		if w == nil {
+			return nil, nil, false
+		}
+
+		conn, err := newConnection(p.address, p.opts...)
+		if err != nil {
+			w.tryDeliver(nil, err)
+			return nil, nil, false
+		}
+
+		conn.pool = p
+		conn.poolID = atomic.AddUint64(&p.nextID, 1)
+		p.conns[conn.poolID] = conn
+
+		return w, conn, true
+	}
+
+	for ctx.Err() == nil {
+		w, conn, ok := wait()
+		if !ok {
+			continue
+		}
+
+		if p.monitor != nil {
+			p.monitor.Event(&event.PoolEvent{
+				Type:         event.ConnectionCreated,
+				Address:      p.address.String(),
+				ConnectionID: conn.poolID,
+			})
+		}
+
+		conn.connect(context.Background())
+		err := conn.wait()
+		if err != nil {
+			_ = p.removeConnection(conn, event.ReasonConnectionErrored)
+			_ = p.closeConnection(conn)
+			w.tryDeliver(nil, err)
```

Comment:
IIUC, `tryDeliver` may return false if `w` was satisfied by a concurrent call to `checkIn`. As a result, `err` is not delivered to any caller of `checkOut`.

In the old behavior of `get`, an error in the inline call to `connect` would return immediately.

I do not see any immediate concern. Callers can still see the `ConnectionClosed` events. But I am noting in case this was unintentional.

Reply:
That's correct, the error here may never be delivered to a caller in the following circumstances:
1. Another `checkIn` or `createConnections` process may deliver a connection or error to this waiting `wantConn`.
2. The `checkOut` call may expire, cancelling this `wantConn`.

That isn't necessarily intentional, it's just better to deliver the first idle connection or error to a `wantConn` vs trying to find a `wantConn` to deliver an error to (i.e. it's a necessary side effect of trying to optimistically deliver connections to `wantConn`s). A "ConnectionClosed: connectionError" event will be published, but won't include the actual error message. Does it make sense to include that error message in the published `ConnectionClosed` event?